src/gui/widget_fpsmeter.c

```diff
@@ -0,0 +1,139 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *   * Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of LCUI nor the names of its contributors may be used
+ *     to endorse or promote products derived from this software without
+ *     specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <assert.h>
+#include <string.h>
+#include <stdlib.h>
+#include <LCUI_Build.h>
+#include <LCUI/LCUI.h>
+#include <LCUI/graph.h>
+#include <LCUI/font/fontlibrary.h>
+#include <LCUI/gui/css_library.h>
+#include <LCUI/gui/css_fontstyle.h>
+#include <LCUI/gui/css_parser.h>
+#include <LCUI/font/textlayer.h>
+#include <LCUI/gui/widget_fpsmeter.h>
+
+#define FPS_METER_TEXT_SIZE_MAX		(256)
+#define FPS_METER_UPDATE_TIME_MS	(1000)
+
+static LCUI_FpsMeterRec self = {FALSE, };
+
+
+static const char *css = CodeToString(
+
+textview {
+	color: rgb(255,255,255);
+	font-size: 16px;
+	text-align: left;
+	z-index: INT_MAX;
+	position: absolute;
+	padding: 20px, 20px 20px 20px;
+	background-color: rgb(60, 60, 60);
+	opacity: 0.5;
+}
+
+);
+
+void LCUI_FpsMeter_Update()
+{
+	if (self.is_enabled == FALSE) {
+		return;
+	}
+
+	char buf[FPS_METER_TEXT_SIZE_MAX];
+	snprintf(buf, FPS_METER_TEXT_SIZE_MAX,
+			"FPS: %ld\n"
+			"Render thread count: %d\n"
+			"Widget update count: %ld\n"
+			"Render count: %ld",
+			self.fps,
+			self.render_thread_count,
+			self.widget_update_count,
+			self.render_count);
+	TextView_SetText(self.widget, buf);
+}
+
+void LCUI_FpsMeter_FrameCount(void)
+{
+	self.frame_count++;
+	if (LCUI_GetTimeDelta(self.last_time) >= FPS_METER_UPDATE_TIME_MS) {
+		self.fps = (self.frame_count * 1000 ) / FPS_METER_UPDATE_TIME_MS;
+		self.frame_count = 0;
+		self.last_time = LCUI_GetTime();
+
+		LCUI_FpsMeter_Update();
+	}
+}
+
+void LCUI_FpsMeter_RenderThreadCount(int count)
```

Comment:
We should make this module provide less public interface, I think only one interface is enough:

```c
void LCUI_EnableFpsMeter(LCUI_BOOL enabled);
```

If FpsMeter is enabled, we can create a timer to get profile data every second andrenders to FpsMeter.

Reply:
You can call the function here to get the data:

https://github.com/lc-soft/LCUI/blob/7c9a6d10f627ab3fc7e17ed45e5982e2b5022ab1/src/main.c#L113-L156

But I suggest you modify it to provide the following public interfaces:

```c
void LCUI_GetProfile(LCUI_Profile profile);

void LCUI_EnableProfile(LCUI_BOOL enable);
```